pytest-embedded/pytest_embedded/plugin.py

```diff
@@ -1273,6 +1298,25 @@ def get_param(item: Function, key: str, default: t.Any = None) -> t.Any:
 
         return item.callspec.params.get(key, default) or default
 
+    @staticmethod
+    def add_app_paths_to_testcases(report_path: str, testcase_app_paths_map: dict, rootdir: str) -> None:
+        with open(report_path, 'r+') as file:
+            tree = ET.parse(file)
+            root = tree.getroot()
+
+            for testcase in root.findall('.//testcase'):
+                testcase_name = testcase.get('name')
+                if testcase_name in testcase_app_paths_map:
+                    relative_paths = {os.path.relpath(path, rootdir) for path in testcase_app_paths_map[testcase_name]}
+                    testcase.set('app_path', ':'.join(relative_paths))
+
+            buffer = StringIO()
```

Comment:
What do you think about reusing this instead of the buffer?
```suggestion
      fw.write(escape_illegal_xml_chars(ET.tostring(root, encoding='unicode')))
```

Reply:
I believe using` ElementTree.write` with a buffer is a safer  approach. After` ET.tostring()` creates the XML string, `escape_illegal_xml_chars()` is applied to clean it up by removing illegal characters. However, if the input to `ET.tostring()` already contains illegal characters, they will appear in the serialized string and need to be removed.